src/eval.c

```diff
@@ -273,16 +280,11 @@ void freeLuaScriptsSync(dict *lua_scripts, list *lua_scripts_lru_list, lua_State
     listRelease(lua_scripts_lru_list);
     lua_close(lua);
 
-#if !defined(USE_LIBC)
-    /* The lua interpreter may hold a lot of memory internally, and lua is
-     * using libc. libc may take a bit longer to return the memory to the OS,
-     * so after lua_close, we call malloc_trim try to purge it earlier.
-     *
-     * We do that only when Redis itself does not use libc. When Lua and Redis
-     * use different allocators, one won't use the fragmentation holes of the
-     * other, and released memory can take a long time until it is returned to
-     * the OS. */
-    zlibc_trim();
+#if defined(USE_JEMALLOC)
+    /* When lua is closed, destroy the previously used private tcache. */
+    void *ud = (global_State*)G(lua)->ud;
+    unsigned int lua_tcache = (unsigned int)(uintptr_t)ud;
+    je_mallctl("tcache.destroy", NULL, NULL, (void *)&lua_tcache, sizeof(unsigned int));
```

Comment:
isn't / shouldn't that tcache shared with functions?
if it is, do we need to re-create it for them?
if it isn't, aren't we missing another call to tcache.destroy?

Reply:
> if it isn't, aren't we missing another call to tcache.destroy?
function does not recreate the lua vm in `function flush`, so there is no need to destroy the tcache.
did we miss the lua vm recreate of function?

Reply:
ok, let me see i didn't miss anything.

functions call createLuaState only once, and ever destroy or re-create it.
that takes care of lua_newstate and tcache.crete.

eval calls it once on startup, and on FLUSH destroys the lua_State and tcache and calls it again to re-create both.

on the other hand, the arena is created on server startup, never destroyed, and unlike the tcache, it is shared between functions and eval.

ohh, i now see it's all explained in the top comment.